libraries/lib-numeric-formats/formatters/BeatsNumericConverterFormatter.cpp

```diff
@@ -0,0 +1,234 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/**********************************************************************
+
+ Audacity: A Digital Audio Editor
+
+ @file BeatsNumericConverterFormatter.cpp
+
+ Dmitry Vedenko
+
+ **********************************************************************/
+#include "BeatsNumericConverterFormatter.h"
+
+#include <algorithm>
+#include <array>
+#include <cmath>
+
+#include "NumericConverterRegistry.h"
+
+#include "SampleCount.h"
+
+namespace
+{
+
+class BeatsFormatter final :
+    public NumericConverterFormatter
+{
+public:
+   BeatsFormatter(
+      double tempo, int upperTimeSignature, int lowerTimeSignature,
+      int fracPart)
+       : mTempo { tempo }
+       , mUpperTimeSignature { upperTimeSignature }
+       , mLowerTimeSignature { lowerTimeSignature }
+       , mFracPart { fracPart }
+   {
+      mFields.reserve(mFracPart > 0 ? 3 : 2);
+
+      // Create "bars" field and digits. We don't really care about frac and base,
+      // Range is assumed to allow 999 bars. 
+      auto& barsField = mFields.emplace_back(NumericField { false, 1, 1000, true });
+      barsField.label = L" " + XO("bar").Translation() + L" ";
+
+      // Beats format is 1 based. For the time point "0" the expected output is "1 bar 1 beat [1]"
+      // For this reason we use (uts + 1) as the "range".
+      // On top of that, we want at least two digits to be shown. Audacity has a bug with the range of 11,
+      // where on digit will be calculated instead of 2.
+
+      auto& beatsField = mFields.emplace_back(
+         NumericField { false, 1, std::max(12, mUpperTimeSignature + 1), true });
+      beatsField.label = L" " + XO("beat").Translation();
+
+      if (mFracPart > 0)
+      {
+         beatsField.label += L" ";
+         // See the reasoning above about the range
+         auto& fracField = mFields.emplace_back(
+            NumericField { false, 1, std::max(12, mFracPart / mLowerTimeSignature + 1), true });
```

Comment:
I notice `false` passed to the `frac` ctor argument, a little surprising, but then I also see it only influences a different override of `ValueToString` that this class doesn't use

Reply:
That leads me to think that maybe `NumericFields` should contain only the commonly needed fields, and the subclasses of `NumericConverterFormatter` add whatever else is special-purpose.

`base` is not needed by this new class.

`range` isn't needed by this class either after `CreateDigitFormatStr` uses it to decide the format string.

`zeropad` isn't needed as a member either -- for EITHER subclasss!  It can just be just an argument of the constructor of `NumericField` and `CreateDigitFormatStr `.

`pos` and `formatStr` are used by each subclass, but there is no common code in the base class that uses them, except initialization in the constructor.  So they too might be removed.

That could shrink `NumericField` to just `digits` and `label` -- which are used in the UI layeer code in NumericTextCtrl.

Then that means the public interface of lib-numeric-formats is simplified.  This structure, which is exposed to NumericTextCtrl, would have less in it needed explanation and documentation.